### PR TITLE
Integration tests, scenario #1

### LINKED ISSUE #1336: SPIKE: How to manage the automated tests
What should be checked periodically:
JVM Heap and CPU usage
Checkpoint acceptance rate
Snapshot alignment check
Exceptions/Errors on the cluster
Data migration pipeline
Rewards (either modifying /balance endpoint to return balance from rewards or keeping track of nodes' balances) - to be defined
Stardust balance (as a part of rewards)

Options: Jenkins

### LINKED ISSUE #1438: Test scenario with sending transactions from single address to multiple nodes

### PR DESCRIPTION
**Pull request by tbekas**
Resolves #1438

### CODE REVIEW

File: .github/workflows/integration_tests.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+          java-version: 1.8
+
+      - name: Decrypt secrets
+        run: integration-tests/decrypt-secrets.sh
+        env:
+          IT_SECRET_PASSPHRASE: ${{ secrets.IT_SECRET_PASSPHRASE }}
+
+      - name: Test
+        run: |
+          sbt integrationTests/test
```
Comment (marcinwadon) [inline .github/workflows/integration_tests.yml:25]: This way you will only check few simple scenarios. Instead, it should spin up an EC2 instance which will be a Test Manager. That instance should run different scenarios, verify them and verify metrics periodically.

File: .github/workflows/integration_tests.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+
+      - name: Decrypt secrets
+        run: integration-tests/decrypt-secrets.sh
+        env:
+          IT_SECRET_PASSPHRASE: ${{ secrets.IT_SECRET_PASSPHRASE }}
+
+      - name: Test
```
Comment (marcinwadon) [inline .github/workflows/integration_tests.yml:23]: There is a missing infrastructure spin up part. It should make a terraform apply or destroy and apply on a given workspace.

File: integration-tests/src/test/resources/application.conf [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# TODO replace it with hosts discovery
+hosts = ["50.18.91.185", "54.241.96.75", "54.153.35.114"]
```
Comment (marcinwadon) [inline integration-tests/src/test/resources/application.conf:2]: IMO it should be taken from the terraform output for the given workspace

File: integration-tests/src/test/resources/application.conf [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+tx-gen {
+    dst = "DAG8y1BQXeVr9n6fh5BZz6KwAs2Ubrjc6VMvzxxw"
+    amount = 1000
+    fee = 10
+    count-total = 100
+    count-per-host = 50
+    key-pair {
```
Comment (marcinwadon) [inline integration-tests/src/test/resources/application.conf:10]: we can assume that all keys will use the same alias/keypass/storepas which we do for all test cases and all testing environments. you just need to provide the path for keys but anyway keys should be taken dynamically from the s3

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+  private val containTheSameElements = (not be empty).and(Matcher { (left: Seq[_]) =>
+    MatchResult(
+      left.forall(_ == left.head),
+      left + " did not contain the same elements",
+      left + " did contain the same elements"
+    )
+  })
+
+  "sent transactions should change src and dst address balances" in {
```
Comment (marcinwadon) [inline integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: a single test scenario should be a separate app which can be run independently.

File: wallet/src/main/scala/org/constellation/wallet/Transaction.scala [hunk 6]
```diff
@@ -63,7 +63,7 @@ object TransactionExt {
   ): Transaction = {
     val lastTxRef =
       prevTx
-        .map(tx => LastTransactionRef(tx.hash, tx.lastTxRef.ordinal + 1))
+        .map(tx => LastTransactionRef(tx.hash, tx.ordinal))
```
Comment (marcinwadon) [inline wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: Why?

**Review by marcinwadon — CHANGES_REQUESTED** [review]

File: wallet/src/main/scala/org/constellation/wallet/Transaction.scala (hunk 6 above)
Comment (tbekas) [reply wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: The main reason is easier mocking.I don't need to mock subobjects, I can do it with just `tx.ordinal should return 123`. The secondary reason is DRY.

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 5 above)
Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: What do you exactly mean by independently? I don't see why this can't be run independently from the rest of the code.

File: integration-tests/src/test/resources/application.conf (hunk 3 above)
Comment (tbekas) [reply integration-tests/src/test/resources/application.conf:2]: I think it's better to take it from loadbalancer. In the same way that any user would do it.

File: integration-tests/src/test/resources/application.conf (hunk 4 above)
Comment (tbekas) [reply integration-tests/src/test/resources/application.conf:10]: Why taking it from s3 would be better than keeping it next to the config? It's versioned and easier to manage if it's in repo.

File: .github/workflows/integration_tests.yml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+name: Integration tests
+
+jobs:
+  test:
+    name: Integration tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@master
```
[the remark this answers is not in the document]
Comment (tbekas) [reply .github/workflows/integration_tests.yml:11]: Yep, you're right. But I would keep the master as the default value, for one-click run.

File: .github/workflows/integration_tests.yml (hunk 1 above)
Comment (tbekas) [reply .github/workflows/integration_tests.yml:25]: Verifying metrics continuously, with alerts is grafana's job. Executing the scenarios can be run on CI, I don't see why dedicated EC2 instance would be needed for that.

File: integration-tests/src/test/resources/application.conf (hunk 4 above)
Comment (marcinwadon) [reply integration-tests/src/test/resources/application.conf:10]: Will you keep 1k or 10k keys in the confit in the repo?

File: .github/workflows/integration_tests.yml (hunk 2 above)
Comment (tbekas) [reply .github/workflows/integration_tests.yml:23]: Yep, that will be a part of the #1211 .

File: wallet/src/main/scala/org/constellation/wallet/Transaction.scala (hunk 6 above)
Comment (marcinwadon) [reply wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: But it's a place where it's defined how to create transaction. It should stay as it is and we shouldn't rely on the implementation part (so the additional method in the tx)

File: integration-tests/src/test/resources/application.conf (hunk 4 above)
Comment (tbekas) [reply integration-tests/src/test/resources/application.conf:10]: Ok, I see the problem. I will add keys to s3 along with the infra code.

File: wallet/src/main/scala/org/constellation/wallet/Transaction.scala (hunk 6 above)
Comment (tbekas) [reply wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: Well in the same line we do rely on the implementation by taking: `tx.hash` instead of `tx.edge.observationEdge.hash`.

File: integration-tests/src/test/resources/application.conf (hunk 4 above)
Comment (marcinwadon) [reply integration-tests/src/test/resources/application.conf:10]: Keys are already there

File: .github/workflows/integration_tests.yml (hunk 2 above)
Comment (marcinwadon) [reply .github/workflows/integration_tests.yml:23]: I'd say that infrastructure should be the first, then manager to spin up scenarios and aggregating results, and then scenarios starting from simpler ones to more complex.

Comment (tbekas) [reply .github/workflows/integration_tests.yml:23]: One does not depend on the other. We just need the URL of the loadbalancer and grafana to run the tests and gather the metrics.

File: .github/workflows/integration_tests.yml (hunk 1 above)
Comment (marcinwadon) [reply .github/workflows/integration_tests.yml:25]: CI cannot run it correctly because tests can last longer. Grafana with prometheus is only a tool for aggregating metrics. It will not show you if the test passed or now, it will not aggregate results and combine them into a summary

File: .github/workflows/integration_tests.yml (hunk 2 above)
Comment (marcinwadon) [reply .github/workflows/integration_tests.yml:23]: Load balancer and grafana depends on the test cluster. Scenarios cannot be run if there is no infrastructure to run on. Tests cannot be executed if there is no cluster to run on and manager to run from.

File: integration-tests/src/test/resources/application.conf (hunk 3 above)
Comment (marcinwadon) [reply integration-tests/src/test/resources/application.conf:2]: It depends on the scenario. Most likely we should pass all nodes to the tests but if scenario require a load balancer, we should use load balancer (but keeping in mind all the nodes which are building a test cluster). i.e. you cannot check node joining based on the load balancer only. it will show you nothing

File: wallet/src/main/scala/org/constellation/wallet/Transaction.scala (hunk 6 above)
Comment (marcinwadon) [reply wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: That's true but we rely on the hash method in the whole codebase in the same way. We assume that there is a `hash` for every class which is signable. For the ordinal I see a reason why it's explicitly defined there but it's up to you :)

File: .github/workflows/integration_tests.yml (hunk 1 above)
Comment (tbekas) [reply .github/workflows/integration_tests.yml:25]: The default timeout on github job is 6 hours according to [this docs](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes). I don't think there's a need for any scenarios that will run longer than that and if there will be such a need I would consider alternatives only then. Regarding metrics, we can simply query for metrics during the test-run.

File: .github/workflows/integration_tests.yml (hunk 2 above)
Comment (tbekas) [reply .github/workflows/integration_tests.yml:23]: They're loosely coupled only via the URL. I left the `TODO` to replace the hardcoded hosts IPs with the loadbalancer URL, and I will do it with the next changes.

File: .github/workflows/integration_tests.yml (hunk 1 above)
Comment (marcinwadon) [reply .github/workflows/integration_tests.yml:25]: we would have to pay more for such long timings on github rather than t2.micro on aws

File: .github/workflows/integration_tests.yml (hunk 2 above)
Comment (marcinwadon) [reply .github/workflows/integration_tests.yml:23]: No. Grafana cannot be started without load balancer and load balancer cannot be started without the initial nodes IPs.
Grafana can workaround load balancer but if we would like to use load balancer, terraform should spin up a test cluster, then the load balancer and then the grafana

File: .github/workflows/integration_tests.yml (hunk 1 above)
Comment (tbekas) [reply .github/workflows/integration_tests.yml:25]: Ok, so let's calculate the price to make the right decision. Keep in mind that `t2.micro` might not be enough, it has only 1 GB of ram and 1 cpu. GitHub runners have [2 cpus and 7 GB of ram](https://docs.github.com/en/free-pro-team@latest/actions/reference/specifications-for-github-hosted-runners#supported-runners-and-hardware-resources).

Let say we want to post 10k transactions to a 3 node cluster. The throughput should be 50 txs * 3 nodes * 2 consensus rounds per minute, it will be 300 txs per minute, so 33 minutes for 10k txs. Running it 4 times a day Mon-Fri will cost us around 2,7k minutes of execution time per month. With the [GitHub team plan](https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions) we have 3k minutes for free and any additional minute cost $0.008. So even if all the other workloads like compile & test on each PR would take 1k minutes a month, it would be still cheaper than the cheapest `t2.micro` which is 9.50$ per month.

So both options are rather inexpensive, but with GitHub actions we have:
- more powerful instances (more cpu and ram, which may be handy for load tests)
- we don't need to manage another piece of infrastructure, that needs to be maintained, secured, etc.

File: .github/workflows/integration_tests.yml (hunk 2 above)
Comment (tbekas) [reply .github/workflows/integration_tests.yml:23]: >No.
That's a 'no' to which part, that they're loosely coupled? We just need node IPs, that's all. It doesn't matter if we take it from the load balancer, DNS, or even a static JSON file from S3. There's no need to run terraform apply before the test (unless we want a throw-away environment, that as far as I remember you were against having).

File: wallet/src/main/scala/org/constellation/wallet/Transaction.scala (hunk 6 above)
Comment (tbekas) [reply wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: I still can't understand why would you advocate for duplicating the method implementation, given the argument:
>it's a place where it's defined how to create transaction

Can other guys give their opinion about this issue @kpudlik  @TheMMaciek? Maybe I'm missing something.

File: .github/workflows/integration_tests.yml (hunk 2 above)
Comment (marcinwadon) [reply .github/workflows/integration_tests.yml:23]: They're loosely coupled only via the URL. You don't need to restart the whole infrastructure but you have to spin up an infrastructure

File: wallet/src/main/scala/org/constellation/wallet/Transaction.scala (hunk 6 above)
Comment (TheMMaciek) [reply wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: I get both points, from one side the definition of ordinal should be set in stone so we should be able to reuse it, from to other relying on something that should be set in stone but is defined somewhere outside of current scope and relying on it feels uncertain. After all I think we are aware enough that it's crucial to not change transaction's logic freely and our team is small so we should catch it if someone fiddles with it. Therefore it's ok for me to reuse ordinal definition.

Comment (marcinwadon) [reply wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: @tbekas - I agree with @TheMMaciek

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 5 above)
Comment (TheMMaciek) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: I think we are kind of pulling in two different directions here. One is sanity checking code base which I think we can do with simple integration tests formulated in such a way. The other is creating an infrastructure for running more complex tests, that can be run arbitrarily in different configurations, for different branches, on demand or automatically. I can see scenario written in scala as above as a part of a test, but I think we also need a part that manages and sets the state for the whole test, and then runs such scala app with a scenario or a set of scenarios. I mean I still think we will need to have the ability to ssh to an instance to for example change firewall, or grep logs to check if something happened and if we can proceed with next steps e.g. In the perfect world it would be great to have such an infrastructure that we can run it with some scenarios with GitHub workflow automatically for example for a PR and with others more complex ones on demand with a GitHub action or as a separate AWS environment. Github jobs seem ok at first glance, but also I'm not sure how would we use it to also spin up the whole infrastructure and if it can be done completely automatically. To summarize I think what we have in this PR is a good start, but we need all the other parts to put it together and then verify if this part is still compatible. For sure I think we need some TestManager instance that would conduct the test scenarios and probably that would require us to use some scripting language to do that.

File: wallet/src/main/scala/org/constellation/wallet/Transaction.scala (hunk 6 above)
Comment (tbekas) [reply wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: So to close the discussion, @TheMMaciek you are in favor of keeping the change, am I reading it right?

Comment (TheMMaciek) [reply wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: Yes :).

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 5 above)
Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: So are you talking about having your own Jenkins-like instance, that would run the job periodically?

@TheMMaciek all of the things you mentioned, like logging via SSH into an instance or manipulating firewall _can_ be done with scenarios run with GitHub CI. I can't imagine a scenario that could be run on Jenkins or a "TestManager" deployed to EC2 instance, that couldn't be run on GitHub CI.

Comment (marcinwadon) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: I think so - and maybe not about running tests periodically but running checks periodically (that's for sure) and tests on demand, or scheduled in some way, or a set of tests one after another. I think the problem is not with a single scenario but with everything combined together. I even thought about a custom CLI/UI tool which will manage tests, make summaries, etc. but it's for future.

tl;dr; you can think about such a scenario:

--- **single scenario but long**:
- Send transactions (**not** via the load balancer) from the same source address to different nodes
- Check if all transactions has been processed (based on the tx amount - this can lasts even few hours)

--- **more complex scenario**:
- Use 10 nodes
- Keep sending transactions (i.e. ~100tx/min) via the load balancer
- Wait for snapshots
- Simulate network issue on 1 node
- Check nodegrid for status
- Stop two nodes one after the other
- Check nodegrid for status
- Reconnect that two nodes
- ...

--- **set of scenarios**:
- Check single tx scenario
- Check multiple txs via load balancer
- Check multiple txs mixed to nodes
- Check blacklisting
- Check joining and dropping
- Check network issues
- Check redownloads
- Periodically check snapshot alignment
- Periodically check majority heights
- ...

Such scenarios/set of scenarios can be long. We should be able to run selected scenarios (not always all of them). Personally I think that it will be really hard to do using only GitHub CI, with no intermediate instance/manager or sth like that.

Comment (TheMMaciek) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: I don't mean periodically. If we checked the branch once there is no need to check it again unless there are changes added I guess. I would for example give a possibility to run the tests, or a selected test scenario on a feature branch so the developer can pre test his changes. If you say that we can do everything with GitHub CI then great! Would we then write a separate Github workflow for every scenario that requires custom/adjusted state or configuration?

Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: So the longest scenario you mentioned:

>     Send transactions (not via the load balancer) from the same source address to different nodes
>     Check if all transactions has been processed (based on the tx amount - this can lasts even few hours)

is the one that's implemented with this change, but currently on a small volume. It sends 100 transactions to 2 nodes and so far it takes only 1 minute. Even if we would go for 10k transactions over 3 nodes it still wouldn't be very long. I wouldn't go for more in this scenario because it's not a performance test, rather a consensus test. For performance tests, we will be likely using different source addresses, so the throughput will be higher, resulting in shorter runs.

Comment (marcinwadon) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: but it's not the only scenario which can take longer time and currently it's no time to think about all possible scenarios and calculate how long they can run. The original ticket has a name:
> Long running cluster tests

and that's the business issue/value we want to be solved. Assuming we have tests which can take hours or days, we need a way to run them.

Simple integration tests are something different than long running tests which we should focus on.

Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: Ok, so for now, I would suggest that we proceed with GitHub CI and if there will be ever a need to run test that will take a week or a month to complete we will go for [self-hosted runner](https://docs.github.com/en/free-pro-team@latest/actions/hosting-your-own-runners/about-self-hosted-runners) on EC2. Is it fine with you?

Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: >Would we then write a separate Github workflow for every scenario that requires custom/adjusted state or configuration?

@TheMMaciek If the test logic could be expressed with scala, then I think we would just add another scenario to `integration-test/src/test/scala` and it will be picked automatically by sbt runner. If we however need to use another tool, then I would just add another step to the existing workflow, instead of creating a new one.

File: integration-tests/src/test/resources/application.conf (hunk 3 above)
Comment (kpudlik) [reply integration-tests/src/test/resources/application.conf:2]: Yeah, sometimes you need to query specific endpoint to for example test joining. Querying load balancer only shows the result of one random node from upstream but I guess you may need to check all the nodes.

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 5 above)
Comment (marcinwadon) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: >>>
Workflow run time - Each workflow run is limited to 72 hours. If a workflow run reaches this limit, the workflow run is cancelled.

>>>
Job queue time - Each job for self-hosted runners can be queued for a maximum of 24 hours. If a self-hosted runner does not start executing the job within this limit, the job is terminated and fails to complete.

>>>
API requests - You can execute up to 1000 API requests in an hour across all actions within a repository. If exceeded, additional API calls will fail, which might cause jobs to fail.

Still you would need a manager for this and somehow schedule jobs on CI. Anyway, what's your idea to schedule a scenario set using Github CI only? How will we be able to specify which scenarios we would like to run? How results will be summarised? IMO that's the things we should focus on before even writing a single scenario.

File: wallet/src/main/scala/org/constellation/wallet/Transaction.scala (hunk 6 above)
Comment (kpudlik) [reply wallet/src/main/scala/org/constellation/wallet/Transaction.scala:66]: Touching that part is risky but I get all your points and it is fine for me to keep that change :)

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 5 above)
Comment (marcinwadon) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: > > Would we then write a separate Github workflow for every scenario that requires custom/adjusted state or configuration?
> 
> @TheMMaciek If the test logic could be expressed with scala (note that there is a scala SSH client and java AWS SDK), then I think we would just add another scenario to `integration-test/src/test/scala` and it will be picked automatically by sbt runner. If we however need to use another tool, then I would just add another step to the existing workflow, instead of creating a new one.

I told you about scenarios for which we will use something different than scala because scala wouldn't be so efficient. Also, we would like to specify which scenarios should be ran, which no. Writing only one github job will end up with a really long single test with almost no business value for us.

Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:56]: >Anyway, what's your idea to schedule a scenario set using Github CI only?

Not sure if I understand the question, but there are [scheduled jobs](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#schedule) on GitHub CI.

>How will we be able to specify which scenarios we would like to run?

By default, I would just run all the scenarios every single time, but if there will be a need to just run some of them, we could categorize them by [tags in ScalaTest](https://www.scalatest.org/user_guide/tagging_your_tests).

>How results will be summarised?

We get the reports from the ScalaTest, or any other testing tool that we would use (all of the testing tools I know, have reporting tool built-in).

File: integration-tests/src/test/resources/application.conf (hunk 3 above)
Comment (tbekas) [reply integration-tests/src/test/resources/application.conf:2]: @kpudlik fully agree. The intention was to use LB only for node discovery, with [/cluster/info endpoint](http://lb.testnet.constellationnetwork.io:9000/cluster/info).

File: .github/workflows/integration_tests.yml (hunk 1 above)
Comment (tbekas) [reply .github/workflows/integration_tests.yml:51]: Resolved, we leave GitHub CI for now.

File: integration-tests/src/test/resources/application.conf (hunk 3 above)
Comment (tbekas) [reply integration-tests/src/test/resources/application.conf]: Resolved, the output now is taken from the terraform.

File: .github/workflows/integration_tests.yml (hunk 2 above)
Comment (tbekas) [reply .github/workflows/integration_tests.yml]: Resolved, infrastructure was added in the same workflow.

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 5 above)
Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala]: Resolved, for now we use `ScalaTest` for reports and for running tagged tests.

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+  private def getLasTxRefs(address: String) = evalRequestForHosts[LastTransactionRef] {
+    _.map(host => GET(Uri.unsafeFromString(s"http://$host:9000/transaction/last-ref/$address")))
+  }
+
+  private def evalRequestForHosts[T](pipe: Pipe[IO, String, IO[Request[IO]]])(implicit ev: EntityDecoder[IO, T]) =
+    Stream
+      .emits(terraform.instanceIps.value)
+      .through(pipe)
+      .evalMap { req =>
+        BlazeClientBuilder[IO](global).resource.use { httpClient =>
```
Comment (marcinwadon) [inline integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:136]: Should it create separate client for every request?

File: terraform/aws/dag [hunk 9]
```diff
@@ -29,17 +27,6 @@ checkHosts() {
     echo -e "${red}HOSTS_FILE was not set${clr}"
     exit 1
   fi
-
-  if [ $CONFIRMED -eq 0 ]; then
```
Comment (marcinwadon) [inline terraform/aws/dag:33]: Please don't remove it

Comment (tbekas) [reply terraform/aws/dag:33]: Then I can add a parameter like `auto-approve` (I need to run it on CI)

Comment (marcinwadon) [reply terraform/aws/dag:33]: Yes, you can add something which will skip checks but checks must be there enabled by default.

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 8 above)
Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:136]: I didn't like that as well, but I was struggling with making clients reusable. I can try again and try to fix it.

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    val targetSrcBalance = srcBalances.head - (config.txGen.amount + config.txGen.fee) * config.txGen.countTotal
+    val targetDstBalance = dstBalances.head + config.txGen.amount * config.txGen.countTotal
+
+    eventually {
+      getBalances(src) should contain only targetSrcBalance
+      getBalances(dst) should contain only targetDstBalance
+    }
+  }
+
+  private def getKeyPair =
```
Comment (TheMMaciek) [inline integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:86]: I wonder if we really need to load up the key from a file? I guess for this test we could just as well just create a random keyPair with keytool. Just food for thought if we really need to load the keys.

Comment (TheMMaciek) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:86]: Probably other more complicated tests will need that.

Comment (marcinwadon) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:86]: We need these keys as these addresses are pre-filled in genesis. But we shouldn't fetch all of them, that's for sure

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 8 above)
Comment (marcinwadon) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:136]: You can create a Kleisli resource I think

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 10 above)
Comment (TheMMaciek) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:86]: Right, forgot about genesis, and that we need some positive balance.

Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:86]: I agree with @TheMMaciek ideally we should generate these keys on the fly as well as genesis that corresponds with those keys. I didn't do that with this change, but I think it wouldn't hurt to have a task for it.

This is an open-source project, and if someone else would like to contribute to it, he wouldn't be able to run these tests without access to our S3 bucket.

Comment (marcinwadon) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:86]: @tbekas Everybody is able to run own cluster, there is no need to have an access to our S3.

Comment (marcinwadon) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:86]: yet, you cannot use autogenerated keys

File: .github/workflows/integration_tests.yml [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        run: |
+          sudo apt-get update
+          sudo apt-get install jq
+
+      - name: Start cluster
+        run: |
+          pushd terraform/aws
+          export EXECUTION_TAG=$(git describe --exact-match --tags 2>/dev/null || git rev-parse --short HEAD)
+          terraform workspace new integration-tests-$EXECUTION_TAG
+          ./dag autostart 3
```
Comment (kpudlik) [inline .github/workflows/integration_tests.yml:40]: I am wondering if it should be always 3 🤔 

But let's eventually change that if we need

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 10 above)
Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala:86]: Those keys aren't public, aren't they? So any contributor would have to generate them on his own, as well as genesis. Lot's of unnecessary manual work, that should be automated.

File: .github/workflows/integration_tests.yml (hunk 11 above)
Comment (tbekas) [reply .github/workflows/integration_tests.yml]: Changed to parameterized value, with default of 3.

File: integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala (hunk 8 above)
Comment (tbekas) [reply integration-tests/src/test/scala/org/constellation/TransactionsProcessingSpec.scala]: I changed it to building a client for a single method call.

File: .github/workflows/integration_tests.yml [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+      - name: Set up Terraform
+        uses: hashicorp/setup-terraform@v1
+        with:
+          terraform_version: 0.12.25
+
+      - name: Install dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install jq
```
Comment (marcinwadon) [inline .github/workflows/integration_tests.yml:39]: Please remember about `haveged` if you want to generate keys on the fly

File: build.sbt [hunk 13]
```diff
@@ -183,7 +192,7 @@ lazy val keytool = (project in file("keytool"))
     buildInfoPackage := "org.constellation.keytool",
     buildInfoOptions ++= Seq(BuildInfoOption.BuildTime, BuildInfoOption.ToMap),
     mainClass := Some("org.constellation.keytool.KeyTool"),
-    libraryDependencies ++= keyToolSharedDependencies
+    libraryDependencies ++= (keyToolSharedDependencies ++ testDependencies ++ fs2Dependencies)
```
Comment (marcinwadon) [inline build.sbt:195]: Why keytool needs test and fs2 dependencies?

File: .github/workflows/integration_tests.yml (hunk 12 above)
Comment (tbekas) [reply .github/workflows/integration_tests.yml:39]: Yes, I remember about this tool. But I left out generating keys for later.

File: build.sbt (hunk 13 above)
Comment (tbekas) [reply build.sbt:195]: Good catch 👍  It's a leftover from draft changes to keytool.

File: terraform/aws/dag [hunk 14]
```diff
@@ -245,6 +246,25 @@ Terraform commands:
     grafana \t\t\t Open grafana
 EOF)
 
+POSITIONAL=()
```
Comment (marcinwadon) [inline terraform/aws/dag:249]: Can you write a line in a README? `terraform/aws/README.md`

Comment (tbekas) [reply terraform/aws/dag:249]: Yes, also you reminded about 2 other things:
1. changelog
2. the `usage` docs inside of the dag script

Comment (marcinwadon) [reply terraform/aws/dag:249]: btw - we don't mark changes in changelog if it's our tooling only and it does not affect nodes

Comment (tbekas) [reply terraform/aws/dag:249]: Ok, good to know that.

**Review by kpudlik — CHANGES_REQUESTED** [review]

**Review by marcinwadon — APPROVED** [review]

**Review by kpudlik — APPROVED** [review]
